Extension/src/LanguageServer/settings.ts

```diff
@@ -996,28 +849,90 @@ export class OtherSettings {
         this.resource = resource;
     }
 
-    public get editorTabSize(): number | undefined { return vscode.workspace.getConfiguration("editor", this.resource).get<number>("tabSize"); }
-    public get editorInsertSpaces(): boolean | undefined { return vscode.workspace.getConfiguration("editor", this.resource).get<boolean>("insertSpaces"); }
-    public get editorAutoClosingBrackets(): string | undefined { return vscode.workspace.getConfiguration("editor", this.resource).get<string>("autoClosingBrackets"); }
-    public get filesEncoding(): string | undefined { return vscode.workspace.getConfiguration("files", { uri: this.resource, languageId: "cpp" }).get<string>("encoding"); }
-    public get filesAssociations(): any { return vscode.workspace.getConfiguration("files").get("associations"); }
-    public set filesAssociations(value: any) {
-        void vscode.workspace.getConfiguration("files").update("associations", value, vscode.ConfigurationTarget.Workspace);
+    private getVSCodeSettingAsString(setting: string, configuration: string, resource: any, defaultString?: string): string | undefined {
+        const fullConfiguration = vscode.workspace.getConfiguration(setting, resource);
```

Comment:
The goal of these helpers not returning `undefined` isn't being achieved with the current implementation. When we provide a default value, we want the return type to not include `undefined`.  You can achieve this with overloads in TypeScript. The syntax is weird, something like this:

```
export function getAsString(input: any): string | undefined;
export function getAsString(input: any, defaultValue: string): string;
export function getAsString(input: any, defaultValue?: string): string | undefined {
    if (isString(input)) {
        return input;
    }
    return defaultValue;
}
```

Reply:
I don't know if it is entirely possible to not return undefined if we want to use the default value obtain programmatically like below:

```
 const config = fullConfiguration.inspect<string>(setting);
```

The default value property of the configuration can possibly be undefined. We could just hardcode our default values to get around this.

What do you think?

Reply:
Nvmd, I figured it out.

Reply:
I'm not sure I follow. Did you try the code I shared? If you supply a value IntelliSense will pick the one that does not use `| undefined` and it works out.

Reply:
Oh, maybe I confused you because my function was named differently and had a different signature because it was an example.